### ISSUE TITLE
rfc1: forked repos versus branches

### ISSUE THREAD

**Issue by garlick** [flux-framework/rfc#127]
In flux-framework/flux-core#1543 we realized we need a "stable branch" for flux-core-0.9 which is getting some use in production while we continue developing unstable new features in flux-core master.

[RFC 1](https://github.com/flux-framework/rfc/blob/master/spec_1.adoc) mandates one branch per github project, that the main project master is the unstable development branch and, and that github repo clones are the mechanism for creating stable branches.  So if we follow this, we'd clone flux-core to flux-core-0.9 and its master would be the first stable production "branch".

This comes from zeromq where they have:

https://github.com/zeromq/libzmq - development
https://github.com/zeromq/zeromq4-1 - 4.1 stable
https://github.com/zeromq/zeromq4-x - 4.0 stable
https://github.com/zeromq/zeromq3-x - 3.0 stable

**Pros**

* repos for stable releases can have different maintainers
* issues that may require a backport to a stable release are not confused with resolved issues in the development repo
* zeromq project has used this workflow for many releases, and generally we have had good success borrowing their processes

**Cons**

* issues and PR's have to be cross referenced using the long form flux-framework/repo#number
* depends on github's organization/project structure to relate the repos
* it's a little surprising to people used to working with "release branches"

The alternative is to amend RFC1 to allow stable release branches in the same repo as development.
This would be the more traditional approach.

What do people think?

**Comment by garlick** [comment]
There is also this:

http://hintjens.com/blog:24

**Comment by grondo** [comment]
Hm, I'm really on the fence. Requirement to move to long from issue referencing is a huge drawback and will that even work to close an issue in the current repository? (I haven't tested it)

A quick check in the zeromq repos shows that they seem to just use local issue numbers, avoid mentioning issues, or do merges complete with conflicts... I guess one strategy is to open issues on stable branches for each backport, which then references the long-form issue, and then in the backport commit(s) reference the local issue?

**Comment by garlick** [comment]
> I guess one strategy is to open issues on stable branches for each backport, which then references the long-form issue, and then in the backport commit(s) reference the local issue?

That makes sense.

I found Hintjens' arguments in the blog somewhat convincing, particularly about isolation and keeping things as simple as possible to avoid confusion and disaster, though maybe "disaster" is too strong a word for what can go wrong.

**Comment by grondo** [comment]
Yes, that blog entry was helpful. However, with each project in the flux-framework organization following this same scheme we are going to have an explosion of forks.

I didn't really buy that forks make things as simple as possible. If you want to backport or forward port fixes you have to have multiple remotes, it requires those extra steps when opening issues and issue renumbering, and it will complicate the release process since URLs to releases have to be slightly edited.

However, I'm willing to try it out.

**Comment by chu11** [comment]
My first thought was, "How do a submit a pull request to a non-master branch?".  I honestly didn't know b/c PRs/merging to master is the default and all the projects I've submitted PRs to on Github just went to master.  I guess I never really worked on a project on Github large enough to have stable branches.

So, it made me think, do we really want to deviate from this?  We could have multiple pull requests that are identical, but each going to different branches.  It's easy to pick the wrong branch from the pull down menu.  Issue discussions about different branches pollute the same repo's issue discussion.  It feels very un-Github like.

Makes me lean to trying the zeromq way.

**Comment by grondo** [comment]
> Issue discussions about different branches pollute the same repo's issue discussion.

On the other hand, it is one project and it would be nice if all issues were reported in one place. zeromq seemed to take this approach and encouraged users to open all issues in the development repos issue tracked, and only open requests for backports in release repos.

I've seen a lot of projects that take pull requests on a "stable" branch vs master. Though I agree it is slightly easier if the stable branches are separate repos, in the real world most of the time you should fix the problem on master then backport to releases when necessary. So the initial fix or development would be a normal PR to master, then a maintainer would do the backport and would presumably know how to use the drop down list.

**Comment by SteVwonder** [comment]
> However, with each project in the flux-framework organization following this same scheme we are going to have an explosion of forks.

Looking at the [zeromq organization page](https://github.com/zeromq?page=1), it seems that the release repos do not really clutter up the org.  I do not know if they manually pinned certain repos higher or if the natural activity on certain repos is what is putting them at the top.  If the latter, one side-effect is that you can instantly tell what the most popular release of zmq is (I'm not really sure if that is a positive or a negative).

Although, on further investigation, it seems this might be because they only have forks for the major releases of libzmq (2.0, 3.0, 4.0, 4.1).  They do not have one for 4.2 or for any of the bindings (pyzmq, pyczmq, etc).
